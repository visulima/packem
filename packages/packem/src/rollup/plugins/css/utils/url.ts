import { isAbsolute, normalize, parse } from "@visulima/path";
import { isRelative } from "@visulima/path/utils";

/**
 * Checks if the URL starts with a tilde followed by a digit, '@', or a letter (case-insensitive)
 * @param url
 * @returns
 */
export const hasModuleSpecifier = (url: string): boolean => /^~[\d@A-Z]/i.test(url);

// handle importing Sass partials in node_modules
// @import ~foo/bar/partial where "partial" has the filename "_partial.scss".
export const getUrlOfPartial = (url: string): string => {
    const { base, dir } = parse(url);

    return dir ? `${normalize(dir)}/_${base}` : `_${base}`;
};

export const normalizeUrl = (url: string): string => {
    if (hasModuleSpecifier(url)) {
        return normalize(url.slice(1));
    }

    if (isAbsolute(url) || isRelative(url)) {
        return normalize(url);
    }

    return `./${normalize(url)}`;
};
