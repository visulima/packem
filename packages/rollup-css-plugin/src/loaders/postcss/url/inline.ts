import liteMime from "mime/lite";

export default (file: string, source: Uint8Array): string => {
    const mime = liteMime.getType(file) ?? "application/octet-stream";
    const data = Buffer.from(source).toString("base64");

    return `data:${mime};base64,${data}`;
};
