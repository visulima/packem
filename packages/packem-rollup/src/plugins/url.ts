/**
 * Modified copy of https://github.com/rollup/plugins/blob/master/packages/url/src/index.js
 *
 * The MIT License (MIT)
 *
 * Copyright (c) 2019 RollupJS Plugin Contributors (https://github.com/rollup/plugins/graphs/contributors)
 */
import crypto from "node:crypto";
import { createReadStream, createWriteStream } from "node:fs";
import { readFile, stat } from "node:fs/promises";

import type { FilterPattern } from "@rollup/pluginutils";
import { createFilter } from "@rollup/pluginutils";
import { ensureDir } from "@visulima/fs";
import { basename, dirname, extname, join, relative } from "@visulima/path";
import mime from "mime";
import type { Plugin } from "rollup";

import { svgEncoder } from "../utils";

const copy = async (source: string, destination: string): Promise<void> => {
    await new Promise((resolve, reject) => {
        const read = createReadStream(source);

        read.on("error", reject);

        const write = createWriteStream(destination);

        write.on("error", reject);
        write.on("finish", () => resolve(undefined));

        read.pipe(write);
    });
};

export interface UrlOptions {
    /**
     * The destination dir to copy assets, usually used to rebase the assets according to HTML files.
     * @type {string}
     */
    destDir?: string;
    /**
     * If false, will prevent files being emitted by this plugin. This is useful for when you are using Rollup to emit both a client-side and server-side bundle.
     * @type {boolean}
     * @default true
     */
    emitFiles: boolean;
    /**
     * A picomatch pattern, or array of patterns, which specifies the files in the build the plugin
     * should _ignore_.
     *
     * By default, no files are ignored.
     * @type {FilterPattern}
     */
    exclude?: FilterPattern;
    /**
     * If emitFiles is true, this option can be used to rename the emitted files. It accepts the following string replacements:
     * [hash] - The hash value of the file's contents
     * [name] - The name of the imported file (without its file extension)
     * [extname] - The extension of the imported file (including the leading .)
     * [dirname] - The parent directory name of the imported file (including trailing /)
     * @type {string}
     * @default [hash][extname]
     */
    fileName: string;
    /**
     * A picomatch pattern, or array of patterns, which specifies the files in the build the plugin
     * should operate on.
     * By default, the png,jpg,jpeg,gif,svg,webp files are targeted.
     * @type {FilterPattern}
     */
    include?: FilterPattern;
    /**
     * The file size limit for inline files.
     * If a file exceeds this limit, it will be copied to the destination folder and the hashed filename will be provided instead.
     * If limit is set to 0 all files will be copied.
     * @type {number}
     * @default 14336 (14kb)
     */
    limit: number;
    /**
     * A string which will be added in front of filenames when they are not inlined but are copied.
     * @type {string}
     */
    publicPath?: string;

    /**
     * When using the [dirname] replacement in fileName, use this directory as the source directory from which to create the file path rather than the parent directory of the imported file. For example:
     * @example
     * ```js
     *    src/path/to/file.js
     *
     *    import png from './image.png';
     *    rollup.config.js
     *
     *    url({
     *      fileName: '[dirname][hash][extname]',
     *      sourceDir: path.join(__dirname, 'src')
     *    });
     *
     *    Emitted File: path/to/image.png
     * ```
     */
    sourceDir?: string;
}

export const urlPlugin = ({
    destDir: destinationDirectory,
    emitFiles,
    exclude,
    fileName,
    include,
    limit,
    publicPath,
    sourceDir: sourceDirectory,
}: UrlOptions): Plugin => {
    const filter = createFilter(include, exclude);
    const copies: Record<string, string> = {};

    return <Plugin>{
        async generateBundle(outputOptions: { dir?: string; file?: string }) {
            if (!emitFiles) {
                return;
            }

            const base = destinationDirectory ?? outputOptions.dir ?? dirname(outputOptions.file ?? "");

            await ensureDir(base);

            await Promise.all(
                Object.keys(copies).map(async (name) => {
                    const output = copies[name] as string;
                    const outputDirectory = join(base, dirname(output));

                    await ensureDir(outputDirectory);
                    await copy(name, join(base, output));
                }),
            );
        },

        async load(id: string) {
            if (!filter(id)) {
                return undefined;
            }

            this.addWatchFile(id);

            const [stats, buffer] = await Promise.all([stat(id), readFile(id)]);

            let data: string;

            if ((limit && stats.size > limit) || limit === 0) {
                const hash = crypto.createHash("sha1").update(buffer).digest("hex").slice(0, 16);
                const extension = extname(id);
                const name = basename(id, extension);
                const relativeDirectory = sourceDirectory ? relative(sourceDirectory, dirname(id)) : basename(dirname(id));

                const outputFileName = fileName
                    .replaceAll("[hash]", hash)
                    .replaceAll("[extname]", extension)
                    .replaceAll("[dirname]", relativeDirectory === "" ? "" : `${relativeDirectory}/`)
                    .replaceAll("[name]", name);

                data = join(publicPath ?? "", outputFileName);

                copies[id] = outputFileName;
            } else {
                const mimetype = mime.getType(id);

                if (mimetype === undefined) {
                    throw new Error(`Could not determine mimetype for ${id}`);
                }

                const isSVG = mimetype === "image/svg+xml";

                data = isSVG ? svgEncoder(buffer) : buffer.toString("base64");

                data = `data:${mimetype};base64,${data}`;
            }

            return `export default "${data}"`;
        },
        name: "packem:url",
    };
};
