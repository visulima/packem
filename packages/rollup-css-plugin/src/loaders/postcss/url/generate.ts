import { getHash } from "@visulima/packem-share/utils";
import { basename, parse } from "@visulima/path";

import { HASH_REGEXP } from "../constants";

export default (placeholder: string, file: string, source: Uint8Array): string => {
    const { base, dir, ext, name } = parse(file);
    const hash = getHash(`${base}:${Buffer.from(source).toString()}`);
    const match = HASH_REGEXP.exec(placeholder);
    const hashLength = match && Number.parseInt(match[1] as string, 10);

    return placeholder
        .replace("[dir]", basename(dir))
        .replace("[name]", name)
        .replace("[extname]", ext)
        .replace(".[ext]", ext)
        .replace("[ext]", ext.slice(1))
        .replace(HASH_REGEXP, hashLength ? hash.slice(0, hashLength) : hash.slice(0, 8));
};
