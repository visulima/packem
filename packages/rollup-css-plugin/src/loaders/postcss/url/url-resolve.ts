import { readFileSync } from "@visulima/fs";

import { resolve } from "../../../utils/resolve";

/** File resolved by URL resolver */
export interface UrlFile {
    /** Absolute path to file */
    from: string;

    /** File source */
    source: Uint8Array;

    /** Original query extracted from the input path */
    urlQuery?: string;
}

/** URL resolver */
export type UrlResolve = (inputUrl: string, baseDirectories: string[]) => Promise<UrlFile>;

export const urlResolve: UrlResolve = async (inputUrl: string, baseDirectories: string[]): Promise<UrlFile> => {
    const urlObject = new URL(inputUrl, "file://");
    const fragmentIdentifier = urlObject.hash ? urlObject.hash.slice(1) : "";
    const url = inputUrl.split("?")[0] ?? "";

    const paths = [url];

    if (url.startsWith("/")) {
        paths.push(`.${url}`);
    }

    if (!url.startsWith("/") && !url.startsWith(".")) {
        paths.push(`./${url}`);
    }

    const from = resolve(paths, { baseDirs: baseDirectories, caller: "URL resolver" });
    const urlQuery = new URLSearchParams(urlObject.search).toString();

    return {
        from,
        source: readFileSync(from, { buffer: true }),
        urlQuery: (urlQuery ? `?${urlQuery}` : "") + (fragmentIdentifier ? `#${fragmentIdentifier}` : ""),
    };
};
