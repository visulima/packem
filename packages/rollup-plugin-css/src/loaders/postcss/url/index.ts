import { basename, dirname, isAbsolute, join, normalize } from "@visulima/path";
import type { Declaration, PluginCreator } from "postcss";
import type { Node, ParsedValue } from "postcss-value-parser";
import valueParser from "postcss-value-parser";

import { mm } from "../../../utils/sourcemap";
import { DATA_URI_REGEXP, FIRST_EXTENSION_REGEXP } from "../constants";
import generateName from "./generate";
import inlineFile from "./inline";
import type { UrlFile, UrlResolve } from "./url-resolve";
import { urlResolve } from "./url-resolve";
import { isDeclWithUrl, walkUrls } from "./utils";

const name = "packem-css-url";
const placeholderHashDefault = "assets/[name]-[hash][extname]";
const placeholderNoHashDefault = "assets/[name][extname]";
const defaultPublicPath = "./assets/";
const defaultAssetDirectory = ".";

const plugin: PluginCreator<UrlOptions> = (userOptions) => {
    const options = {
        alias: {},
        assetDir: defaultAssetDirectory,
        inline: false,
        resolve: urlResolve,
        ...userOptions,
    };
    const placeholder = options.hash ?? true ? typeof options.hash === "string" ? options.hash : placeholderHashDefault : placeholderNoHashDefault;

    return {
        async Once(css, { result }) {
            if (!css.source?.input.file) {
                return;
            }

            const { file } = css.source.input;

            const map = mm(css.source.input.map?.text ?? undefined)
                .resolve(dirname(file))
                .toConsumer();

            const urlList: {
                baseDirs: Set<string>;
                decl: Declaration;
                node: Node;
                parsed: ParsedValue;
                url: string;
            }[] = [];

            const imported = new Set(result.messages.filter((message) => message.type === "dependency").map((message) => message.file as string));

            css.walkDecls((decl) => {
                if (!isDeclWithUrl(decl)) {
                    return;
                }

                const parsed = valueParser(decl.value);

                walkUrls(parsed, (url, node) => {
                    // Resolve aliases
                    for (const [from, to] of Object.entries(options.alias)) {
                        if (url !== from && !url.startsWith(`${from}/`)) {
                            continue;
                        }

                        // eslint-disable-next-line no-param-reassign
                        url = normalize(to) + url.slice(from.length);
                    }

                    // Empty URL
                    if (!node || url.length === 0) {
                        decl.warn(result, `Empty URL in \`${decl.toString()}\``);

                        return;
                    }

                    // Skip Data URI
                    if (DATA_URI_REGEXP.test(url)) {
                        return;
                    }

                    // Skip Web URLs
                    if (!isAbsolute(url)) {
                        try {
                            // eslint-disable-next-line no-new
                            new URL(url);

                            return;
                        } catch {
                            // Is not a Web URL, continuing
                        }
                    }

                    const baseDirectories = new Set<string>();

                    // Use PostCSS imports
                    if (decl.source?.input.file && imported.has(decl.source.input.file)) {
                        baseDirectories.add(dirname(decl.source.input.file));
                    }

                    // Use SourceMap
                    if (decl.source?.start) {
                        const pos = decl.source.start;
                        const realPos = map?.originalPositionFor(pos);
                        const basedir = realPos?.source && dirname(realPos.source);

                        if (basedir) {
                            baseDirectories.add(normalize(basedir));
                        }
                    }

                    // Use current file
                    baseDirectories.add(dirname(file));

                    urlList.push({ baseDirs: baseDirectories, decl, node, parsed, url });
                });
            });

            const usedNames = new Map<string, string>();

            for await (const { baseDirs, decl, node, parsed, url } of urlList) {
                let resolved: UrlFile | undefined;

                try {
                    if (!resolved) {
                        resolved = await options.resolve(url, [...baseDirs]);
                    }
                } catch {
                    /* noop */
                }

                if (!resolved) {
                    decl.warn(result, `Unresolved URL \`${url}\` in \`${decl.toString()}\``);

                    continue;
                }

                const { from, source, urlQuery } = resolved;

                if (!(source instanceof Uint8Array) || typeof from !== "string") {
                    decl.warn(result, `Incorrectly resolved URL \`${url}\` in \`${decl.toString()}\``);

                    continue;
                }

                result.messages.push({ file: from, plugin: name, type: "dependency" });

                if (options.inline) {
                    node.type = "string";
                    node.value = inlineFile(from, source);
                } else {
                    const unsafeTo = normalize(generateName(placeholder, from, source));
                    let to = unsafeTo;

                    // Avoid file overrides
                    const hasExtension = FIRST_EXTENSION_REGEXP.test(unsafeTo);

                    // eslint-disable-next-line no-plusplus
                    for (let index = 1; usedNames.has(to) && usedNames.get(to) !== from; index++) {
                        to = hasExtension ? unsafeTo.replace(FIRST_EXTENSION_REGEXP, `${String(index)}$1`) : `${unsafeTo}${String(index)}`;
                    }

                    usedNames.set(to, from);

                    const resolvedPublicPath
                        = typeof options.publicPath === "string"
                            ? options.publicPath + (/[/\\]$/.test(options.publicPath) ? "" : "/") + basename(to)
                            : `${defaultPublicPath}${basename(to)}`;

                    node.type = "string";
                    node.value = typeof options.publicPath === "function" ? options.publicPath(node.value, resolvedPublicPath, file) : resolvedPublicPath;

                    if (urlQuery) {
                        node.value += urlQuery;
                    }

                    if (typeof options.assetDir === "string") {
                        to = join(options.assetDir, to);
                    } else if (typeof options.assetDir === "function") {
                        to = options.assetDir(from, to, file);
                    }

                    result.messages.push({ plugin: name, source, to, type: "asset" });
                }

                decl.value = parsed.toString();
            }
        },
        postcssPlugin: name,
    };
};

plugin.postcss = true;

/** URL handler options */
export interface UrlOptions {
    /**
     * Aliases for URL paths.
     * Overrides the global `alias` option.
     * - ex.: `{"foo":"bar"}`
     */
    alias?: Record<string, string>;

    /**
     * Directory path for outputted CSS assets,
     * which is not included into resulting URL
     * @default "."
     */
    assetDir?: string | ((original: string, resolved: string, file: string) => string);

    /**
     * Enable/disable name generation with hash for outputted CSS assets
     * or provide your own placeholder with the following blocks:
     * - `[extname]`: The file extension of the asset including a leading dot, e.g. `.png`.
     * - `[ext]`: The file extension without a leading dot, e.g. `png`.
     * - `[hash(:&lt;num>)]`: A hash based on the name and content of the asset (with optional length).
     * - `[name]`: The file name of the asset excluding any extension.
     *
     * Forward slashes / can be used to place files in sub-directories.
     * @default "assets/[name]-[hash][extname]" ("assets/[name][extname]" if false)
     */
    hash?: boolean | string;

    /**
     * Inline files instead of copying
     * @default true for `inject` mode, otherwise false
     */
    inline?: boolean;

    /**
     * Public Path for URLs in CSS files
     * @default "./"
     */
    publicPath?: string | ((original: string, resolved: string, file: string) => string);

    /**
     * Provide custom resolver for URLs
     * in place of the default one
     */
    resolve?: UrlResolve;
}

export default plugin;
