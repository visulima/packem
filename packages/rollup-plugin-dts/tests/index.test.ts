import path from "node:path";
import { fileURLToPath } from "node:url";

import { rolldownBuild } from "@sxzz/test-utils";
import { expect } from "vitest";

import { dts } from "../src/index.js";

const dirname = path.dirname(fileURLToPath(import.meta.url));

it("basic", async () => {
    const { snapshot } = await rolldownBuild(
        path.resolve(dirname, "fixtures/basic.ts"),
        [dts()],
    );

    expect(snapshot).toMatchSnapshot();
});

it("tsx", async () => {
    const { snapshot } = await rolldownBuild(
        path.resolve(dirname, "fixtures/tsx.tsx"),
        [dts()],
    );

    expect(snapshot).toMatchSnapshot();
});

it("resolve dependencies", async () => {
    const { snapshot } = await rolldownBuild(
        path.resolve(dirname, "fixtures/resolve-dep.ts"),
        [
            dts({
                emitDtsOnly: true,
                isolatedDeclarations: true,
                resolve: ["get-tsconfig"],
            }),
        ],
    );

    expect(snapshot).contain("type TsConfigResult");
    expect(snapshot).not.contain("node_modules/rolldown");
});

it("resolve dts", async () => {
    const { snapshot } = await rolldownBuild(
        path.resolve(dirname, "fixtures/resolve-dts/index.ts"),
        [dts()],
    );

    expect(snapshot).matchSnapshot();
});

// Test alias mapping based on rolldown input option
it("input alias", async () => {
    const root = path.resolve(dirname, "fixtures/alias");
    const { chunks, snapshot } = await rolldownBuild(
        null!,
        [dts({ emitDtsOnly: false })],
        {
            cwd: root,
            // A mapping from output chunk names to input files. This mapping should
            // be used in both JS and DTS outputs.
            input: {
                output1: "input1.ts",
                "output2/index": "input2.ts",
            },
        },
    );
    const fileNames = chunks.map((chunk) => chunk.fileName).sort();

    // The JS output and DTS output should have the same structure
    expect(fileNames).toContain("output1.d.ts");
    expect(fileNames).toContain("output1.js");
    expect(fileNames).toContain("output2/index.d.ts");
    expect(fileNames).toContain("output2/index.js");

    expect(snapshot).toMatchSnapshot();
});

it("isolated declaration error", async () => {
    const error = await rolldownBuild(
        path.resolve(dirname, "fixtures/isolated-decl-error.ts"),
        [
            dts({
                emitDtsOnly: true,
                isolatedDeclarations: true,
            }),
        ],
    ).catch((error: any) => error);

    expect(String(error)).toContain(
        `Function must have an explicit return type annotation with --isolatedDeclarations.`,
    );
    expect(String(error)).toContain(`export function fn() {`);
});

it("paths", async () => {
    const root = path.resolve(dirname, "fixtures/paths");
    const { snapshot } = await rolldownBuild(path.resolve(root, "index.ts"), [
        dts({
            emitDtsOnly: true,
            isolatedDeclarations: true,
            tsconfig: path.resolve(root, "tsconfig.json"),
        }),
    ]);

    expect(snapshot).toMatchSnapshot();
});

it("tree-shaking", async () => {
    const { snapshot } = await rolldownBuild(
        path.resolve(dirname, "fixtures/tree-shaking/index.ts"),
        [
            dts(),
            {
                name: "external-node",
                resolveId(id) {
                    if (id.startsWith("node:")) { return { external: true, id, moduleSideEffects: false }; }
                },
            },
        ],
        { treeshake: true },
    );

    expect(snapshot).matchSnapshot();
});

it("dts input", async () => {
    const { snapshot } = await rolldownBuild(null!, [dts({ dtsInput: true })], {
        input: {
            index: path.resolve(dirname, "fixtures/dts-input.d.ts"),
        },
    });

    expect(snapshot).toMatchSnapshot();
});

it("same-name output", async () => {
    const { chunks } = await rolldownBuild(
        [path.resolve(dirname, "fixtures/same-name/index.ts")],
        [dts()],
        {},
        { entryFileNames: "foo.d.ts", preserveModules: true },
    );

    expect(chunks.every((chunk) => chunk.fileName.endsWith(".d.ts"))).toBe(true);
});
