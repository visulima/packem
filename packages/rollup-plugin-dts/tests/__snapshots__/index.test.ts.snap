// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic 1`] = `
"// basic.d.ts
//#region tests/fixtures/basic.d.ts
declare const foo: number;
type SomeType<T> = T;
type FooType = string;
interface Interface {}
declare function fn(arg0: SomeType<FooType>, opt: Interface): void;
declare enum Enum {
  A = 0,
  B = 1,
  C = 2,
}
declare class Cls {
  foo: string;
  fn(e: Enum): void;
}
//#endregion
export { Cls, Enum, fn, foo };
// basic.js
//#region tests/fixtures/basic.ts
const foo = 42;
function fn(arg0, opt) {}
let Enum = /* @__PURE__ */ function(Enum$1) {
	Enum$1[Enum$1["A"] = 0] = "A";
	Enum$1[Enum$1["B"] = 1] = "B";
	Enum$1[Enum$1["C"] = 2] = "C";
	return Enum$1;
}({});
var Cls = class {
	foo;
	fn(e) {}
};

//#endregion
export { Cls, Enum, fn, foo };"
`;

exports[`dts input 1`] = `
"// index.d.ts
//#region tests/fixtures/dts-input.d.ts
declare class Foo {}
//#endregion
export { Foo };"
`;

exports[`input alias 1`] = `
"// input2-CcmD5Ns-.d.ts
//#region shared.d.ts
interface Shared {
  shared: string;
}
declare const shared: Shared;
//#endregion
//#region input2.d.ts
interface Input2 extends Shared {
  input2: string;
}
declare const input2: Input2;
//#endregion
export { Input2, input2 };
// input2-LCiYuoaZ.js
//#region shared.ts
const shared = { shared: "shared" };

//#endregion
//#region input2.ts
const input2 = {
	...shared,
	input2: "input2"
};

//#endregion
export { input2 };
// output1.d.ts
import { Input2 } from "./input2-CcmD5Ns-.js";

//#region input1.d.ts
interface Input1 extends Input2 {
  input1: string;
}
declare const input1: Input1;
//#endregion
export { Input1, input1 };
// output1.js
import { input2 } from "./input2-LCiYuoaZ.js";

//#region input1.ts
const input1 = {
	...input2,
	input1: "input1"
};

//#endregion
export { input1 };
// output2/index.d.ts
import { Input2, input2 } from "../input2-CcmD5Ns-.js";
export { Input2, input2 };
// output2/index.js
import { input2 } from "../input2-LCiYuoaZ.js";

export { input2 };"
`;

exports[`paths 1`] = `
"// index.d.ts
//#region tests/fixtures/paths/mod.d.ts
declare const foo = 42;
//#endregion
export { foo };"
`;

exports[`resolve dts 1`] = `
"// index.d.ts
//#region tests/fixtures/resolve-dts/mod.d.ts
type Foo = string;
//#endregion
export { Foo };
// index.js
"
`;

exports[`tree-shaking 1`] = `
"// index.d.ts
//#region tests/fixtures/tree-shaking/mod.d.ts

type A = string;
//#endregion
export { A };
// index.js
"
`;

exports[`tsx 1`] = `
"// tsx.d.ts
import { JSX } from "solid-js";

//#region tests/fixtures/tsx.d.ts
declare function createComponent(): JSX.HTMLAttributes<HTMLElement>;
//#endregion
export { createComponent };
// tsx.js
import { jsx } from "react/jsx-runtime";

//#region tests/fixtures/tsx.tsx
function createComponent() {
	return /* @__PURE__ */ jsx("div", {});
}

//#endregion
export { createComponent };"
`;
